src/jit/gentree.cpp

```diff
@@ -15161,52 +15161,101 @@ void Compiler::gtExtractSideEffList(GenTree*  expr,
+                    if (node->OperIs(GT_ADDR) && node->gtGetOp1()->OperIsIndir() &&
+                        (node->gtGetOp1()->TypeGet() == TYP_STRUCT))
                     {
-                        printf("Keep the GT_ADDR and GT_IND together:\n");
-                    }
+#ifdef DEBUG
+                        if (m_compiler->verbose)
+                        {
+                            printf("Keep the GT_ADDR and GT_IND together:\n");
+                        }
 #endif
+                        m_sideEffects.Push(node);
+                        return Compiler::WALK_SKIP_SUBTREES;
+                    }
+                }
+
+                // Generally all GT_CALL nodes are considered to have side-effects.
+                // So if we get here it must be a helper call that we decided it does
+                // not have side effects that we needed to keep.
+                assert(!node->OperIs(GT_CALL) || (node->AsCall()->gtCallType == CT_HELPER));
+            }
+
+            if ((m_flags & GTF_IS_IN_CSE) != 0)
+            {
+                // If we're doing CSE we also need to extract CSE definitions and unmark removed CSE uses.
+                if (VisitCSEDefUse(node))
+                {
+                    // This node must be a CSE def and it can not be removed.
+                    assert(IS_CSE_DEF(node->gtCSEnum));
+
                     m_sideEffects.Push(node);
                     return Compiler::WALK_SKIP_SUBTREES;
                 }
+
+                // The existence of CSE definitions (and uses) is not propagated up the tree like side
+                // effects are. We need to continue visiting the tree as if it has side effects.
+                treeHasSideEffects = true;
+            }
+
+            return treeHasSideEffects ? Compiler::WALK_CONTINUE : Compiler::WALK_SKIP_SUBTREES;
+        }
+
+    private:
+        bool VisitCSEDefUse(GenTree* node)
+        {
+            assert(m_compiler->optValnumCSE_phase);
+
+            if (!m_compiler->optUnmarkCSE(node))
+            {
+#ifdef DEBUG
+                if (m_compiler->verbose)
+                {
+                    printf("Preserving the CSE def #%02d at ", GET_CSE_INDEX(node->gtCSEnum));
+                    m_compiler->printTreeID(node);
+                    printf(" because it is nested inside a CSE use\n");
+                }
+#endif // DEBUG
+
+                return true;
             }
 
-            // Generally all GT_CALL nodes are considered to have side-effects.
-            // So if we get here it must be a helper call that we decided it does
-            // not have side effects that we needed to keep.
-            assert(!node->OperIs(GT_CALL) || (node->AsCall()->gtCallType == CT_HELPER));
+            // The call to optUnmarkCSE(node) should have cleared any CSE info.
+            assert(!IS_CSE_INDEX(node->gtCSEnum));
 
-            return Compiler::WALK_CONTINUE;
+            if (node->OperIs(GT_LCL_VAR))
```

Comment:
@AndyAyersMS Broken ref count update in existing CSE code? I see no reason to decrement the ref count for `GT_LCL_VAR` and ignore `GT_LCL_FLD`.

Reply:
Oops, the existing CSE code is actually here: 
https://github.com/dotnet/coreclr/blob/d38774d80e07436eadc28f7255a8c9f8edb1e570/src/jit/optcse.cpp#L281-L307

Reply:
My ambition is to remove incremental ref counting entirely.